### ISSUE TITLE
Treesitter: syntax highlighting for command line, window

### ISSUE THREAD

**Issue by clason** [neovim/neovim#26346]
### Problem

The command line does not have syntax highlighting, despite having a shiny new Vimscript parser at home. Also, enabling treesitter highlighting for vimscript breaks/disables the legacy parser for the command window (`q:`).

### Expected behavior

Hook a treesitter highlighter into the command line (and window) so we can have rainbow goodness (including highlighting of embedded Lua). This requires extending the `TSHighlighter` with a new mode that parses each line independently as its own tree (or writing a custom `TSLineHighlighter`).

This will also allow us to remove the custom (unmaintained) vimscript parser used for `<c-r>=` expression highlighting (although that will require some trickery to parse the whole line as Vim expression rather than statement).

**Comment by justinmk** [comment]
Should we also eliminate the [":help expr-highlight" groups](https://github.com/neovim/neovim/blob/988b472d902e6116353fc2ae605bed32a37e6391/src/nvim/highlight_group.c#L463-L645) (`NvimPlainAssignment`, etc.)? These seem over-specific and I would think we want to use our core highlight groups instead.

**Comment by rudiejd** [comment]
I've been poking around with this a little bit today, and I found that if you set `syntax=vim` in the cmdwin, you get what looks like Treesitter syntax highlighting. This seems like a similar issue to https://github.com/folke/noice.nvim/blob/92433164e2f7118d4122c7674c3834d9511722ba/lua/noice/text/syntax.lua#L12.   I made an autocommand for my config:

```lua
vim.api.nvim_create_autocmd('CmdwinEnter', {
  pattern = '[:>]',
  desc = 'If the treesitter vim parser is installed, set the syntax again to get highlighting in the command window',
  group = vim.api.nvim_create_augroup('cmdwin_syntax', {}),
  callback = function ()
    local is_loadable, _ = pcall(vim.treesitter.language.add, 'vim')
    if is_loadable then
      vim.cmd('set syntax=vim')
    end
  end
})
```

I'm going to keep digging into a less hacky patch for both of these

**Comment by Sam-programs** [comment]
How can treesitter be used in the c code?

**Comment by rudiejd** [comment]
@Sam-programs my idea was to turn the cmdline into a buffer, expose that buffer to lua, then start treesitter in the lua code. I haven't gotten very far with it though

**Comment by Sam-programs** [comment]
@rudiejd I made a [plugin](https://github.com/Sam-programs/cmdline-hl.nvim) for it, I used the string parser not a buffer thought it was too hard to refresh the buffer's highlights.
>  start treesitter in the lua code

I have to call treesitter from lua?

**Comment by clason** [comment]
I think the best approach here would be to use `nvim_ui_attach()` (or `vim.ui.attach` from Lua) to hook into `ext_cmdline` so you can have a (floating window) buffer as the actual commandline. Obviously, this would require adding support for that.

**Comment by LamprosPitsillos** [comment]
> I think the best approach here would be to use `nvim_ui_attach()` (or `vim.ui.attach` from Lua) to hook into `ext_cmdline` so you can have a (floating window) buffer as the actual commandline. Obviously, this would require adding support for that.

So in the end the cmdline bar will become a floating window?

**Comment by clason** [comment]
That's exactly what noice.nvim is already doing; it just needs to be a lot more robust (and less fancy) to be acceptable in core.

**Comment by clason** [comment]
> Note: Well `/` or `?` can be highlighted in Vim script treesitter parser.

More like the `regex` parser. That would require bundling it, though, and I'm not 100% convinced that's a good idea.

**Comment by LamprosPitsillos** [comment]
> > Note: Well `/` or `?` can be highlighted in Vim script treesitter parser.
> 
> More like the `regex` parser. That would require bundling it, though, and I'm not 100% convinced that's a good idea.

I mean regex has been in vim since the beginning, I don't see it as that bad improving it's parsing and highlighting in core.

>That's exactly what noice.nvim is already doing; it just needs to be a lot more robust (and less fancy) to be acceptable in core.

What will become of the `cmdheight` and other cmdline specific options and functions?

**Comment by clason** [comment]
> I mean regex has been in vim since the beginning, I don't see it as that bad improving it's parsing and highlighting in core.

I mean the regex treesitter parser.

> What will become of the cmdheight and other cmdline specific options and functions?

I don't know; hopefully they'll become obsolete.

But these are minor details and not relevant to the core of this issue; please keep comments focused to avoid yet another "80 hidden comments" issue where valuable comments get lost. If you have idle questions, better ask on https://matrix.to/#/#neovim-dev:matrix.org.

**Comment by LamprosPitsillos** [comment]
> Obviously a lot of dead code would need to be removed and probably edge cases considered.

Plus the cmdwindow that needs a special implementation, something like `TSHighlightLine` as mentioned above 

I guess that could be a later PR

**Comment by rudiejd** [comment]
@luukvbaal thank you for the starter code!! I had similar settings in C (dummy buffer, new window) but was missing all of the float config. I was also way overthinking starting treesitter. I'll try this out and see how it goes.

**Comment by rudiejd** [comment]
@luukvbaal oh nice! I interpreted it as the former. feel free, I won't have much time to make progress on this

**Comment by justinmk** [comment]
> to align on whether we actually prefer a C implementation over ext_cmdline.

No objection to a C impl if these are true:

- it's more straightforward in the near-term (not tons of complicated code)
- it isn't a "one-way door" (shifting to a different approach in the future won't require disentangling fragile special-cases spread around the C codebase).

**Comment by justinmk** [comment]
> But I don't think it's wort the effort if ext_cmdline is seen as the way forward

No preference. If this feature gets delivered in C we wouldn't spend time rewriting it as a `ext_cmdline` handler unless there was a strong reason to do so.

**Comment by luukvbaal** [comment]
Hmm replacing the cmdline with a floating window is doable but quite cumbersome. The interactions with the current cmdline and the message grid (on which it is drawn) are quite involved and identifying/replacing all of them requires some work.

I wasn't aware before but while working on this I found out that we already have a way to color the command line lol. I now wonder if it will be more straightforward to just use a scratch buffer with the cmdline contents, attach treesitter (that part I have finished), extract the applied highlights, and still draw them on the message grid through `color_cmdline()`.

Though it seems the only way to do that is through the experimental `g:Nvim_color_cmdline`. If we can deprecate this color callback, along with the VimL/expression parser by replacing the cmdline with a colored floating window, and if that would be preferred, I can continue with resolving display issues. Any thoughts?

**Comment by LamprosPitsillos** [comment]
I believe in the long run it would be more beneficial for the bar ( or even the whole message grid )  to become a float which from what i understand would imply that it would also be a buffer , thus allowing anything that you can do in a buffer ( like snippets, treesitter etc ).

Or maybe even natively move the bar around, since it's a float i can see that happening.

**Comment by luukvbaal** [comment]
Yes, the floating window approach brings possibilities for further niceties (matchparen?)/floating window customizations. 

A floating position completely unattached from the message grid might actually make things easier, but I doubt we want that by default.

**Comment by clason** [comment]
> hough it seems the only way to do that is through the experimental g:Nvim_color_cmdline.

This has to go. The point of this issue is to _replace_ this adhoc parser with treesitter.

**Comment by luukvbaal** [comment]
> This has to go. The point of this issue is to _replace_ this adhoc parser with treesitter.

Sure, but the highlight callback mechanism is a separate component to the parser. One which I thought might be used to apply treesitter highlighting if keeping the cmdline on the message grid would be desirable. 

But I think decoupling it from the message grid is doable, and worthwhile so I continued working on it. Just need to make sure things are redrawn as expected, which gets quite complicated with a moving cmdline row and scrolled messages etc.

Here is my progress so far: https://github.com/luukvbaal/neovim/commit/d92b0570f42d850af10f454247a270f4d790de3e (won't be working on it for a while, currently on the road on vacation).

**Comment by clason** [comment]
Yes, the less special treatment, the better. (This is why I originally suggested `ext_cmdline`, which noice has successfully made use of -- albeit with hacks on top of hacks.)

This is also an opportunity for revisiting `cmdheight=0`, which has been a constant source of issues.

**Comment by luukvbaal** [comment]
Yeah completely eliminating the `!ui_has(kCmdLine)` path also seems quite attractive if a more minimal version of noice were implemented in core. My (uneducated) understanding was that noice requires quite some hacks currently. That and the fact that **ext**_cmdline seems inappropriate for a core component prompted me to explore a C implementation.

**Comment by clason** [comment]
I'm not against that, mind you; especially if it's easier or more robust. We'll also have to consider that GUIs may still want to hijack `ext_cmdline`.

But we _do_ want to make `vim.ui_attach()` a first-class consumer of `ext_foo`, so if we can get rid of the need for hacks (in C), so much the better. This is not a trivial change and thus not part of the 0.10 milestone.

**Comment by tom-gora** [comment]
> [@rudiejd](https://github.com/rudiejd) I made a [plugin](https://github.com/Sam-programs/cmdline-hl.nvim) for it, I used the string parser not a buffer thought it was too hard to refresh the buffer's highlights.
> 
> > start treesitter in the lua code
> 
> I have to call treesitter from lua? Like this ?
> 
>   NLUA_EXEC_STATIC("return vim.something()", args, &err);

LOVE  YOU! Was able to root out noice for the one feature I liked it which custom cmdline rendering, plus I can configure syntax hl in one place? Shut up and take my  my  github star! How is this not a popular plugin?

**Comment by clason** [comment]
Closed by #27855 (with ongoing cleanup work). Thanks @luukvbaal !
